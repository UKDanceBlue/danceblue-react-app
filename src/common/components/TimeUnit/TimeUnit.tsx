import { Text } from "native-base";
import { View } from "react-native";

const validUnits = [
  "seconds", "minutes", "hours", "days", "months", "years"
] as const;

/**
 * A label for a unit of time
 * @param props.value The value of time
 * @param props.unit The unit of time (plural)
 */
const TimeUnit = ({
  value,
  unit,
}: {
  value: number | undefined;
  unit: typeof validUnits[number];
}) => {
  if (!validUnits.includes(unit)) {
    throw new Error(`Invalid unit: ${unit}`);
  }

  return (
    <View style={{ alignItems: "center", paddingLeft: 7, paddingRight: 7 }}>
      <Text style={{
        color: "white",
        fontSize: 40,
        fontWeight: "bold",
      }}>{Math.max(0, value ?? 0)}</Text>
      <Text style={{ fontSize: 20, color: "white" }}>
        {(value ?? 0) === 1 ? unit.substring(0, unit.length - 1) : unit}
      </Text>
    </View>
  );
};

export default TimeUnit;
