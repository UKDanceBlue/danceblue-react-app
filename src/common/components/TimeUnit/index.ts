import TimeUnit from './TimeUnit';
export default TimeUnit;
