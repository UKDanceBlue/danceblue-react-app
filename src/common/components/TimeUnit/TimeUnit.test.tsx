import { render } from "@testing-library/react-native";

import TimeUnit from ".";

describe("<TimeUnit />", () => {
  it("renders correctly with a singular unit", () => {
    const tree = render(<TimeUnit value={1} unit={"seconds"} />).toJSON() as unknown;

    expect(tree).toMatchSnapshot();
  });
  it("renders correctly with a plural unit", () => {
    const tree = render(<TimeUnit value={10} unit={"seconds"} />).toJSON() as unknown;

    expect(tree).toMatchSnapshot();
  });

  it("renders 0 with a negative value", () => {
    const tree = render(<TimeUnit value={-1} unit={"seconds"} />);

    const valueElement = tree.queryAllByText("0");
    expect(valueElement).toHaveLength(1);

    const unitElement = tree.queryAllByText("seconds");
    expect(unitElement).toHaveLength(1);
  });

  it("renders 0 with a nullish value", () => {
    // @ts-expect-error Testing nullish values
    const tree = render(<TimeUnit value={null} unit={"seconds"} />);

    const valueElement = tree.queryAllByText("0");
    expect(valueElement).toHaveLength(1);

    const unitElement = tree.queryAllByText("seconds");
    expect(unitElement).toHaveLength(1);
  });

  it("throws when passed an invalid type", () => {
    const mockedConsoleError = jest.spyOn(console, "error").mockImplementation(() => undefined);

    const invalidUnit = "an invalid unit";
    // @ts-expect-error Testing an invalid unit
    expect(() => render(<TimeUnit value={1} unit={invalidUnit} />)).toThrow(`Invalid unit: ${invalidUnit}`);

    mockedConsoleError.mockRestore();
  });
});
