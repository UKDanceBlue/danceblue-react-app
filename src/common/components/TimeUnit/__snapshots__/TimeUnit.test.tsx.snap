// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TimeUnit /> renders correctly with a plural unit 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    dataSet={Object {}}
    style={
      Array [
        Object {},
        Object {
          "alignItems": "center",
          "paddingLeft": 7,
          "paddingRight": 7,
        },
      ]
    }
  >
    <Text
      dataSet={Object {}}
      style={
        Array [
          Object {
            "backgroundColor": undefined,
            "color": "#171717",
            "fontFamily": undefined,
            "fontSize": 14,
            "fontStyle": "normal",
            "fontWeight": "400",
            "letterSpacing": 0,
            "lineHeight": 21,
            "textDecorationLine": undefined,
          },
          Object {
            "color": "white",
            "fontSize": 40,
            "fontWeight": "bold",
          },
        ]
      }
    >
      10
    </Text>
    <Text
      dataSet={Object {}}
      style={
        Array [
          Object {
            "backgroundColor": undefined,
            "color": "#171717",
            "fontFamily": undefined,
            "fontSize": 14,
            "fontStyle": "normal",
            "fontWeight": "400",
            "letterSpacing": 0,
            "lineHeight": 21,
            "textDecorationLine": undefined,
          },
          Object {
            "color": "white",
            "fontSize": 20,
          },
        ]
      }
    >
      seconds
    </Text>
  </View>
</RNCSafeAreaProvider>
`;

exports[`<TimeUnit /> renders correctly with a singular unit 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    dataSet={Object {}}
    style={
      Array [
        Object {},
        Object {
          "alignItems": "center",
          "paddingLeft": 7,
          "paddingRight": 7,
        },
      ]
    }
  >
    <Text
      dataSet={Object {}}
      style={
        Array [
          Object {
            "backgroundColor": undefined,
            "color": "#171717",
            "fontFamily": undefined,
            "fontSize": 14,
            "fontStyle": "normal",
            "fontWeight": "400",
            "letterSpacing": 0,
            "lineHeight": 21,
            "textDecorationLine": undefined,
          },
          Object {
            "color": "white",
            "fontSize": 40,
            "fontWeight": "bold",
          },
        ]
      }
    >
      1
    </Text>
    <Text
      dataSet={Object {}}
      style={
        Array [
          Object {
            "backgroundColor": undefined,
            "color": "#171717",
            "fontFamily": undefined,
            "fontSize": 14,
            "fontStyle": "normal",
            "fontWeight": "400",
            "letterSpacing": 0,
            "lineHeight": 21,
            "textDecorationLine": undefined,
          },
          Object {
            "color": "white",
            "fontSize": 20,
          },
        ]
      }
    >
      second
    </Text>
  </View>
</RNCSafeAreaProvider>
`;
